Objective-C/CBLFragment.m

```diff
@@ -0,0 +1,139 @@
+//
+//  CBLFragment.m
+//  CouchbaseLite
+//
+//  Created by Pasin Suriyentrakorn on 4/21/17.
+//  Copyright © 2017 Couchbase. All rights reserved.
+//
+
+#import "CBLFragment.h"
+#import "CBLData.h"
+#import "CBLDocument+Internal.h"
+#import "CBLJSON.h"
+
+@implementation CBLFragment {
+    id _value;
+    id _parent;
+    id _parentKey;
+    NSUInteger _index;
+}
+
+
+- /* internal */ (instancetype) initWithValue: (id)value parent: (id)parent parentKey: (id)parentKey {
+    self = [super initWithValue: value];
+    if (self) {
+        _value = value;
+        _parent = parent;
+        _parentKey = parentKey;
+    }
+    return self;
+}
+
+
+#pragma mark - SUBSCRIPTION
+
+
+- (CBLFragment*) objectForKeyedSubscript: (NSString*)key {
+    if ([_value conformsToProtocol: @protocol(CBLDictionary)])
+        return [_value objectForKeyedSubscript: key];
+    return [[CBLFragment alloc] initWithValue: nil parent: nil parentKey: nil];
+}
+
+
+- (CBLFragment*) objectAtIndexedSubscript: (NSUInteger)index {
+    if ([_value conformsToProtocol: @protocol(CBLArray)])
+        return [_value objectAtIndexedSubscript: index];
+    return [[CBLFragment alloc] initWithValue: nil parent: nil parentKey: nil];
+}
+
+
+#pragma mark - SET
+
+
+- (void) setValue: (NSObject*)value {
```

Comment:
It just uses the parent to set the value at the key or index. We can chat about this.